### PR TITLE
Server: Jobs as child actors

### PR DESCRIPTION
**Pull request by pfcoperez**
This PR changes the way a `ServerActor` reacts to new query messages:

Instead of running launching the query to run within the Spark Cluster and block the actor until the job has finished, it creates a child actor, `JobActor`, which is the one launching the job and managing it result  thus allowing the `ServerActor` to run the next queries at its queue.

This new architecture enables the easy implementation of services such as jobs cancellations and state checks. Besides, it adds a job supervision policy which restarts failed jobs. This policy can be easily configurable.

Cool features which could be now easily added:

- Query with a number of tries and timeout.
- Add a job status API.
- Add a job management (cancel, re-play, ...) API.
- Add a query cache.
- Enable resources quota.
- Improve the driver server use pattern.

### CODE REVIEW

File: server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  object Events {
+    case class JobCompleted() extends JobEvent
+    case class JobFailed(err: Throwable) extends JobEvent
+  }
+
+  object Commands {
+    trait JobCommand
+    case class GetJobStatus()
```
Comment (darroyocazorla) [inline server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala]: case object?

Comment (pfcoperez) [inline server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala]: I fully agree, good catch!

File: server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      self ! JobStarted()
+      val rows = if(command.retrieveColumnNames)
+        df.asInstanceOf[XDDataFrame].flattenedCollect() //TODO: Replace this cast by an implicit conversion
+      else df.collect()
+
+      // Sends the result to the requester
+      requester ! SuccessfulQueryResult(command.queryId, rows, df.schema)
+
+      // Complete the job lifecycle
+      self ! JobCompleted
```
Comment (miguel0afd) [inline server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala]: JobCompleted is being sent 2 times. The one from the "onComplete" should be enough.

File: server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  private def receive(status: JobStatus): Receive = {
+    case GetJobStatus =>
+      sender ! status
+    case JobStarted() if status == Starting =>
+      context.become(receive(Running))
+    case event @ JobFailed(e) if sender == self && (Seq(Starting, Running) contains status) =>
+      context.become(receive(Failed))
+      context.parent ! event
+      requester ! ErrorResult(command.queryId, e.getMessage, Some(new Exception(e.getMessage)))
+      throw e //Let It Crash: It'll be managed by its supervisor
```
Comment (miguel0afd) [inline server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala]: Cool! :muscle:

**Comment by miguel0afd** [comment]
Nice work! :+1:

File: server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala (hunk 3 above)
Comment (jjlopezm) [inline server/src/main/scala/com/stratio/crossdata/server/actors/JobActor.scala]: +11111111111111

**Comment by jjlopezm** [comment]
+1

**Comment by compae** [comment]
Nice! I want be as you when I am old

**Pull request merged**